### ISSUE TITLE
Mixed Precision (float16) lead to nan loss (soft_dice_score is inf)

### ISSUE THREAD

**Issue by yohann84L** [qubvel/segmentation_models.pytorch#463]
#### Issue description
It seems when using mixed precision the function `soft_dice_score` return inf because of the intersection between output and target.
However, I'm not sure if I'm doing something wrong or if it's a real issue.

#### Steps to reproduce the issue
For an image of size `(512, 512, 3)`, output and target we'll be of the size 512*512 = 262 144, which is larger than 16-bit float (6.55e4)
```python
import torch
# In float16
output = torch.zeros((262144), dtype=torch.float16) + 1
target = torch.zeros((262144), dtype=torch.float16) + 1
intersection = torch.sum(output * target)
>>> tensor(inf, dtype=torch.float16)

# In float32
output = torch.zeros((262144), dtype=torch.float32) + 1
target = torch.zeros((262144), dtype=torch.float32) + 1
intersection = torch.sum(output * target)
>>> tensor(262144.)
```

Thus, the function `soft_dice_score` return inf which lead to a nan dice loss.
